### PR TITLE
Initial implementation of the 'Dapp approval' mechanism

### PR DESCRIPTION
**Pull request by alex-forshtat-tbk**
1. Move 'checkSig' to the GsnUtils
2. Rename 'sig' to 'approval' in contracts code
3. Add 'approval' to the 'accept_relayed_call'
4. Implement a sample signature check in SampleRecipient

### CODE REVIEW

File: contracts/RelayRecipient.sol [hunk 1]
```diff
@@ -92,8 +92,10 @@ contract RelayRecipient is RelayRecipientApi {
 	 * 			the contract may check the method-id for valid methods
 	 *  @param gas_price - the gas price for this transaction
 	 *  @param transaction_fee - the relay compensation (in %) for this transaction
+	 *  @param approval - first 65 bytes are reserved for the sender's signature, and the rest is
```
Comment (drortirosh) [inline contracts/RelayRecipient.sol]: first 65 bytes are checked by the RelayHub
(relayRecipient doesn't have to check them)

File: contracts/SampleRecipient.sol [hunk 2]
```diff
@@ -46,12 +47,24 @@ contract SampleRecipient is RelayRecipient, Ownable {
...
+        
+        // this is an example of how the dapp can provide an offchain approval to a transaction
+        if (approval.length == 65){
+            // No owner signature given - proceed as usual (for existing tests)
+            return 0;
+        }
+
+        // extract owner sig from all approval bytes
+        bytes memory ownerSig = LibBytes.slice(approval, 65, 130);
+        if (!GsnUtils.checkSig(owner(), keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", keccak256("I approve"))), ownerSig)) {
```
Comment (drortirosh) [inline contracts/SampleRecipient.sol]: add sender to the "I approve sender="

File: src/js/relayclient/RelayClient.js [hunk 3]
```diff
@@ -321,6 +322,20 @@ class RelayClient {
                 signature = await getTransactionSignature(this.web3, options.from, hash);
             }
 
+            if (typeof options.approveFunction === "function"){
+                approval = await options.approveFunction(
```
Comment (drortirosh) [inline src/js/relayclient/RelayClient.js]: named params, in `({})`

File: src/js/relayclient/RelayClient.js [hunk 4]
```diff
@@ -321,6 +322,20 @@ class RelayClient {
...
+                    options.to,
+                    encodedFunctionCall,
+                    options.txfee,
+                    gasPrice,
+                    gasLimit,
+                    nonce,
+                    relayHub._address,
+                    relayAddress)
+            }
+            signature += approval
```
Comment (drortirosh) [inline src/js/relayclient/RelayClient.js]: move into 'if approveFunction' block

File: test/relay_client_test.js [hunk 5]
```diff
@@ -68,41 +68,51 @@ contract('RelayClient', function (accounts) {
...
-            allowed_relay_nonce_gap: 0,
-            verbose: process.env.DEBUG
-        }
+    var func = async function (/*from, to, tx, txfee, gas_price, gas_limit, nonce, relay_hub_address, relay_address*/) {
+        let sign = await utils.getTransactionSignature(web3, accounts[0], web3.utils.sha3("I approve"));
+        return sign.slice(2);
+    }
+    var arr = [null, func]
+    arr.forEach(approveFunction => {
+        it("should send transaction to a relay and receive a response (approveFunction=" + typeof approveFunction + ")", async function () {
```
Comment (drortirosh) [inline test/relay_client_test.js]: `( typeof approveFunction == 'function' ) ? "with" : "without" ) + " approveFunction"`

File: contracts/RelayRecipient.sol (hunk 1 above)
Comment (alex-forshtat-tbk) [reply contracts/RelayRecipient.sol]: Done

File: contracts/SampleRecipient.sol (hunk 2 above)
Comment (alex-forshtat-tbk) [reply contracts/SampleRecipient.sol]: Done

File: src/js/relayclient/RelayClient.js (hunk 3 above)
Comment (alex-forshtat-tbk) [reply src/js/relayclient/RelayClient.js]: Done

File: src/js/relayclient/RelayClient.js (hunk 4 above)
Comment (alex-forshtat-tbk) [reply src/js/relayclient/RelayClient.js]: Done

File: test/relay_client_test.js (hunk 5 above)
Comment (alex-forshtat-tbk) [reply test/relay_client_test.js]: Done

**Pull request merged**